relay-server/src/actors/events.rs

```diff
@@ -644,41 +703,59 @@ impl EventProcessor {
             // event id. To be defensive, we always overwrite to ensure consistency.
             event.id = Annotated::new(event_id);
         } else {
-            // If we have an envelope without event at this point, we are done with processing. This
-            // envelope only contains attachments or user reports. We should not run filters or
-            // apply rate limits.
+            // Processing mode requires an event payload beyond this point.
+            if_processing!({
+                return Err(ProcessingError::NoEventPayload);
+            });
+
+            // If we're here, then there is no event item, but an item that creates an event
+            // placeholder payload in processing mode. Since we don't have an event, pass the
+            // envelope to the upstream Relay immediately.
             log::trace!("no event for envelope, skipping processing");
-            return Ok(ProcessEnvelopeResponse { envelope });
+            return Ok(RateLimits::new());
         }
 
-        if_processing! {
-            self.store_process_event(&mut event, &envelope, &message.project_state)?;
+        // Initialize empty rate limits. Only in processing mode, the rate limiter is checked,
+        // otherwise these limits remain empty;
+        let mut rate_limits = RateLimits::new();
+        rate_limits = rate_limits;
+
+        if_processing!({
+            self.store_process_event(&mut event, &envelope, &project_state)?;
+
+            // Run rate limiting after normalizing the event and running all filters. If the event
+            // is dropped or filtered for a different reason before that, it should not count
+            // against quotas.
+            let category = utils::data_category_from_event(get_event_type(&event));
+            let scope = project_state.get_scoping(envelope.meta());
+            rate_limits = self.enforce_quotas(envelope, &project_state, &scope, Some(category))?;
+
+            // Bail out if the event has been removed by probing the returned rate limits for the
+            // event category. If the event has not been rate limited, we can continue ingestion,
+            // since rate limited items have been removed from the envelope.
+            if rate_limits.check(scope.item(category)).is_limited() {
+                return Ok(rate_limits);
```

Comment:
Note that we do *not* return an error here so that the project can still be updated with the rate limit.